sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -85,7 +85,14 @@ class QueryExecution(
     prepareForExecution(sparkPlan)
   }
 
-  /** Internal version of the RDD. Avoids copies and has no schema */
+  /**
+   * Internal version of the RDD. Avoids copies and has no schema.
+   * Note for callers: Spark may apply various optimization including reusing object: this means
+   * the row is valid only for the iteration it is retrieved. You should avoid storing row and
+   * accessing after iteration. (Calling `collect()` is one of known bad usage.)
+   * If you want to store these rows into collection, please apply some converter or copy row
+   * which produces new object per iteration.
+   */
```

Comment:
@HeartSaVioR Should we point the users to dataset.rdd method where the conversion is already applied ?

Reply:
Yeah that's good suggestion for end users (not Spark developers). Will add.

Reply:
BTW, I don't think it's an API though .. technically we don't have to worry about end users.